[WIP] Add Tokenizer for MyT5 Model

tomlimi opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #31260 


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

tomlimi committed:
Initial commit for MyT5 model

tomlimi committed:
custom implementation of MyT5 tokenizer, unused files deleted

tomlimi committed:
unittest for myt5 tokenizer

amyeroberts commented:
cc @ArthurZucker @itazap

tomlimi committed:
upadate of import structure and style

tomlimi committed:
removed remmanents of MyT5Config

tomlimi committed:
fixed docstrings

tomlimi committed:
Merge remote-tracking branch 'upstream/main' into add_myt5

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31286). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ArthurZucker commented:
Alright! cc @itazap if you can do a first pass!

itazap commented on docs/source/en/model_doc/myt5.md:
```diff
@@ -0,0 +1,48 @@
+
+This model was contributed by [INSERT YOUR HF USERNAME HERE](https://huggingface.co/<INSERT YOUR HF USERNAME HERE>).
+The original code can be found [here](<INSERT LINK TO GITHUB REPO HERE>).
+
```
add inserts!

itazap commented on src/transformers/models/myt5/convert_myt5_original_tf_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,59 @@
+logging.set_verbosity_info()
+
+
+def convert_tf_checkpoint_to_pytorch(tf_checkpoint_path, config_file, pytorch_dump_path):
```
could add a 'Copied from'

itazap commented on src/transformers/models/myt5/tokenization_myt5.py:
```diff
@@ -0,0 +1,347 @@
+        string = bstring.decode("utf-8", errors="ignore")
+        return string
+
+    # MyT5Tokenizer has no vocab file
```
remove # MyT5Tokenizer has no vocab file

itazap commented on src/transformers/models/myt5/tokenization_myt5.py:
```diff
@@ -0,0 +1,347 @@
+VOCAB_FILES_NAMES = {"vocab_file": "byte_maps.json"}
+
+
+class ByteRewriter:
```
add docs & docstring, and maybe a test of this functionality?

itazap commented on tests/models/myt5/test_tokenization_myt5.py:
```diff
@@ -0,0 +1,51 @@
+from transformers import MyT5Tokenizer
+
+
+class MyT5TokenizationTest(unittest.TestCase):
```
test any edges cases, special tokens, and bytewriter cases

itazap reviewed (commented):
@tomlimi Nice, looks clean to me!! :hugs: :hugs: I added a few comments regarding docs, Copied from, and perhaps a couple of edge cases to test to flex the ByteWriter class

tomlimi committed:
Merge remote-tracking branch 'upstream/main' into add_myt5

tomlimi committed:
Updates after review: filled documentaion file, new docstrings and tests added

tomlimi commented:
Sorry for late reply, I pushed the changes addressing the comments in the review. Can this PR be still re-opened?
Thank you!

itazap commented:
Thanks for applying the feedback! Looks good, please update with changes from main and address the minor CI failures (+ ruff reformat)! Otherwise LGTM 🤗

tomlimi committed:
Fixed code style issues

tomlimi committed:
Merge remote-tracking branch 'upstream/main' into add_myt5

tomlimi committed:
fixed copied from to refer to function

tomlimi commented:
I fixed most of CLI errors, unfortunately the issue with run_tests still persitsts. It raises the error, saying that tested Tomlim/myt5-base is not a valid model handle, even though it is public at (I also had this problem with earlier version from June 6). Could it be caused by the way unittest class is declared?

```
E   OSError: Tomlim/myt5-base is not a local folder and is not a valid model identifier listed on 'https://huggingface.co/models'
```

tomlimi committed:
Merge remote-tracking branch 'upstream/main' into add_myt5

tomlimi committed:
updated loading myt5 tokenizer in tests, added sample byte map file to fixtures

tomlimi commented:
I fixed the issue with loading the model by changing the `MyT5TokenizationTest` class, and all the added tests remained unchanged. I also added a sample "vocabulary" file to `fixtures` so the tokenizer can be loaded from it. 

I see that one test fails in CI, but it is in `DPTModelTest` and is unrelated to the changed code.

itazap commented on tests/models/myt5/test_tokenization_myt5.py:
```diff
@@ -0,0 +1,196 @@
+    return bytes_to_hex(bytes(line, "utf-8"), sep)
+
+
+class TestByteRewriter(unittest.TestCase):
```
Nice! 💯

itazap reviewed (approved):
Thanks for applying the feedback anad adding ByteRewriter tests, looks great! 🚀 I reran the CI tests & all pass 🟢 
@ArthurZucker any further comments / test suggestions?

tomlimi commented:
Thank you from prompt reply! Please let me know if there are any other changes or tests that should be added before merging!

itazap commented:
There might be a couple of minor conflicts to resolve!

tomlimi committed:
resolved conflicts with main

tomlimi commented:
The conflict was caused by another model being added in the meantime. I resolved it and merged the new changes from the main 🙂.

tomlimi commented:
@itazap @ArthurZucker can this PR be merged? I recently presented MyT5 at ACL and expect increased interest in the model (tokenizer) in the upcoming time. Thanks!

ArthurZucker commented on tests/models/myt5/test_tokenization_myt5.py:
```diff
@@ -0,0 +1,196 @@
+
+    def setUp(self):
+        super().setUp()
+        tokenizer = MyT5Tokenizer(SAMPLE_VOCAB)
+        tokenizer.save_pretrained(self.tmpdirname)
```
let's use the base instead!

tomlimi replied:
I had tried to load it from the hub, but it did not work for me (I described an error in a comment a month ago).

ArthurZucker replied:
Why not just use ` MyT5Tokenizer.from_pretrained("Tomlim/myt5-base")`

ArthurZucker reviewed (commented):
In general LGTM! The code is clean I have not read the paper, but sounds interesting! 
Would just like to have a short description of how this work in the myt5.md! 
Left kind of an open question about added tokens and potential slowdowns

tomlimi committed:
Merge remote-tracking branch 'upstream/main' into add_myt5

tomlimi committed:
changes after review

tomlimi committed:
removed redundant copied from

tomlimi commented:
Thank you for the review, @ArthurZucker. I added a short description of the method in the model doc and added the missing "Copied from statements". To address the other points:

1. Added tokens conversion to bytes. Actually, in `myt5_tokenizer` we do not use `added_tokens` . Morphemes are rewritten using the original vocabulary of 256 bytes,  MYTE is a rewriting method that does not require additional tokens. I kept the support of `added_tokens` in case someone would like to extend the vocabulary in their instance of the tokenizer.

2. `byte_map.json` in the hub. It is already there ([Tomlim/myt5-base](https://huggingface.co/Tomlim/myt5-base)). In the testing class, I used the vocab file from `fixtures` as it is done, e.g., in T5. I had tried to load it from the hub, but it did not work for me (I described an error in a comment a month ago).

tomlimi committed:
removed redundant copied from

tomlimi commented on tests/fixtures/byte_maps.json:
It is already there ([Tomlim/myt5-base](https://huggingface.co/Tomlim/myt5-base)). In the testing class, I used the vocab file from fixtures as it is done, e.g., in T5.

ArthurZucker commented on src/transformers/models/myt5/tokenization_myt5.py:
```diff
@@ -0,0 +1,373 @@
+                out_tokens.append(token)
+
+        out_tokens = self.morphological_decode(out_tokens)
+        for token in out_tokens:
+            if token in set(self.added_tokens_decoder.values()) | set(self.added_tokens_encoder):
+                bstring += bytes(token, "utf-8")
+            else:
+                bstring += bytes.fromhex(token)
```
Yep got it, but you are re-computing the set of added tokens decoder and added tokens encoder for token in out tokens. Can be pretty slow

ArthurZucker replied:
IMO the best to keep supporting added tokens is to overwrite `_add_tokens` to apply morpheme on the content but otherwise let's just update the computation

ArthurZucker commented on tests/fixtures/byte_maps.json:
let's remove this, again we should be able to load the tokenizer from pretrained

tomlimi committed:
optimalization and loading model from hf

tomlimi committed:
Merge remote-tracking branch 'upstream/main' into add_myt5

ArthurZucker commented:
I think this looks good, let's rebase and merge WDYT? 🤗

tomlimi committed:
Merge remote-tracking branch 'upstream/main' into add_myt5

tomlimi committed:
[run_slow] myt5

HuggingFaceDocBuilderDev commented:
Hey! 🤗 Thanks for your contribution to the `transformers` library!

Before merging this pull request, slow tests CI should be triggered. To enable this:
- Add the `run-slow` label to the PR
- When your PR is ready for merge and all reviewers' comments have been addressed, push an empty commit with the command `[run-slow]` followed by a comma separated list of all the models to be tested, i.e. `[run_slow] model_to_test_1, model_to_test_2`
  - If the pull request affects a lot of models, put at most 10 models in the commit message
- A `transformers` maintainer will then approve the workflow to start the tests

(For maintainers) The documentation for slow tests CI on PRs is [here](https://www.notion.so/huggingface2/CI-for-pull-requests-8335bd9217d24d1e8ed1d3a016f39804).

tomlimi commented:
Thank you! I merged the updates from the main to the branch and pushed an empty commit with `run-slow` as instructed by the bot 🤖. I see some tests are failing but they are not related to the added code.

tomlimi committed:
[run-slow] myt5

HuggingFaceDocBuilderDev commented:
Hey! 🤗 Thanks for your contribution to the `transformers` library!

Before merging this pull request, slow tests CI should be triggered. To enable this:
- Add the `run-slow` label to the PR
- When your PR is ready for merge and all reviewers' comments have been addressed, push an empty commit with the command `[run-slow]` followed by a comma separated list of all the models to be tested, i.e. `[run_slow] model_to_test_1, model_to_test_2`
  - If the pull request affects a lot of models, put at most 10 models in the commit message
- A `transformers` maintainer will then approve the workflow to start the tests

(For maintainers) The documentation for slow tests CI on PRs is [here](https://www.notion.so/huggingface2/CI-for-pull-requests-8335bd9217d24d1e8ed1d3a016f39804).

ArthurZucker reviewed (commented):
Let's remove this and merge

tomlimi committed:
Updated en documentation for myt5

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

ArthurZucker commented:
Oups, `Tomlim/myt5-base` seems private! Could you make it public? CIs are failing because of this!

tomlimi commented:
The model is already public. I had this issue with test suite claiming that it was private before, see this [comment](https://github.com/huggingface/transformers/pull/31286#issuecomment-2301458035) from August. 

That was way I tried solving it bt adding tokenization files to `fixtures`, which I have already removed.

ArthurZucker commented:
Okay I'll have a look!

ArthurZucker commented:
<img width="684" alt="image" src="https://github.com/user-attachments/assets/fb48aed8-1d09-4fc3-8783-36a7b2867989">
Works for me, merging!

ArthurZucker commented:
Thanks for the contribution! 🤗
